### DISCUSSION TITLE
Double Check

### DISCUSSION THREAD

**Discussion by salrodri** [khoih-prog/WebServer_WT32_ETH01#14] [Q&A]
Hi, I programed this board WT32_ETH01 with ESP3D. https://github.com/luc-github/ESP3D

ESP3D is used to add Wi-Fi to a 3D printer. (Serial/Wi-Fi bridge using configurable port 8888)

I was able to flash this board with thee EP3D firmware and it seems to be working correctly.

My question is.
Once I connect the WT32_ETH01 to my router through the port RJ45. How do I know if the board is connecting trough Ethernet instead of internet?

thank you,

**Comment by khoih-prog** [comment]
**It must be WiFi** as the library currently has no code to support `WT32_ETH01` Ethernet.
You can post an `enhancement request` issue or modify by yourself that good and powerful [**ESP3D** library](https://github.com/luc-github/ESP3D) to support `WT32_ETH01`

I'll possibly port that library later, if having time and enough interests, to support many other powerful boards, such as `WT32_ETH01`, `Portenta_H7`, `Teensy 4.1`, `ESP8266 using W5500/ENC28J60 Ethenet`, `RP2040 / RP2040W`, etc. to work with my current **AsyncWebServer** libraries

1. [**AsyncWebServer_WT32_ETH01**](https://github.com/khoih-prog/AsyncWebServer_WT32_ETH01)
2. [**Portenta_H7_AsyncWebServer**](https://github.com/khoih-prog/Portenta_H7_AsyncWebServer)
3. [**AsyncWebServer_Teensy41**](https://github.com/khoih-prog/AsyncWebServer_Teensy41)
4. [**AsyncWebServer_Ethernet**](https://github.com/khoih-prog/AsyncWebServer_Ethernet)
5. [**AsyncWebServer_RP2040W**](https://github.com/khoih-prog/AsyncWebServer_RP2040W)

**Comment by salrodri** [comment]
Thank you for answering my question.
I will post an enhancement request.
I do not feel confident enough to do it by myself.

Alternatively could I use a raspberry pi with octoprint.
Would this be possible?

Would this work over ethernet?

I am working on a project that requests to control a 3D printer wireless
and also have a port for ethernet.

How would you do it?

Please point me in the right direction.

thank you,

**Comment by khoih-prog** [comment]
> Alternatively could I use a raspberry pi with octoprint. Would this be possible?

Anything is possible, just requiring knowledge, preparation and time. Harder to program for full `RPi` (RPi 3, RPi 4) than `RPI-pico`

> Would this work over ethernet?

I don't see why not, just programming the correct way.

> I am working on a project that requests to control a 3D printer wireless
> and also have a port for ethernet.

**Simplest and easiest way is to use `WT32_ETH01` as you correctly selected.**

> How would you do it?
> Please point me in the right direction.

Sorry I don't have time for this now. Please post the request on Arduino Forum, etc. But I doubt anybody would have time to help you with this complex project.

Good Luck,

**Comment by salrodri** [comment]
Understood.
I greatly appreciate your feedback.

thank you,

**Comment by khoih-prog** [comment]
It seems that in v3.0+ has some support for `WT32_ETH01` in [**ESP32 ETH support**](https://github.com/luc-github/ESP3D/tree/3.0/esp3d/src/modules/ethernet)
and

https://github.com/luc-github/ESP3D/blob/102192528dbad3fac9235663fe93464b42f300ed/esp3d/configuration.h#L100-L108

```
/* Use WiFi
* Enable wifi communications
*/
#define WIFI_FEATURE

/* Use Ethernet
* Enable ethernet communications
*/
//#define ETH_FEATURE
```

You have to read and verify if it's working first before posting the enhancement request.

**Comment by salrodri** [comment]
I really appreciate it!!
I will look into it today.

Thank you very much!!

**Comment by salrodri** [comment]
This is awesome !!
I can also add bluetooth !!

**Comment by khoih-prog** [comment]
Great. Please share your experience (modifications, settings, success, etc.) here with other interested people, including me.

**Comment by salrodri** [comment]
Of course.
I would be doing more testing. then will post my results and of course give
you credit.

thank you Mr. Khoih,

**Comment by khoih-prog** [comment]
### To help you to work with ESP32 v2.0.5+

Compile for ESP8266 is OK.

To use `WT32_ETH01`, you better use with ESP32 core 2.0.0+, such as current core `v2.0.5`

But for ESP32, using latest core v2.0.5 + [**ESP3D v3.0.0-alpha3**](https://github.com/luc-github/ESP3D/releases/tag/v3.0.0-alpha3), there are many compile errors

1. Bug in [**arduinoWebSockets**](https://github.com/Links2004/arduinoWebSockets) using even `ESPAsyncTCP library` for ESP32

https://github.com/Links2004/arduinoWebSockets/blob/323592f622e0ec8f9ce1f995c5777d9bbaaae1ec/src/WebSockets.h#L134-L147

```cpp
#if defined(ESP8266)
  #include <ESP8266WiFi.h>
#elif defined(ESP32)
  #include <WiFi.h>
  #include <WiFiClientSecure.h>
  #define SSL_AXTLS
#elif defined(ESP31B)
  #include <ESP31BWiFi.h>
#else
  #error "network type ESP8266 ASYNC only possible on the ESP mcu!"
#endif

#include <ESPAsyncTCP.h>
#include <ESPAsyncTCPbuffer.h>
```

2. 

```
/home/kh/Arduino/libraries/ESP3D-3.0.0-alpha3/esp3d/src/modules/http/handles/handle-description_xml.cpp: In static member function 'static void HTTP_Server::handle_SSDP()':
/home/kh/Arduino/libraries/ESP3D-3.0.0-alpha3/esp3d/src/modules/http/handles/handle-description_xml.cpp:40:48: error: 'class SSDPClass' has no member named 'getSchema'; did you mean 'schema'?
         _webserver->send(200, "text/xml", SSDP.getSchema());
                                                ^~~~~~~~~
                                                schema
```

---


To fix you can use my forked [**ESP3D_Printer branch 3.0**](https://github.com/khoih-prog/ESP3D_Printer/tree/3.0), which fixed those bugs.

#### Fix ESP32 compile errors
1. http module
2. websockets => using [WebSockets_Generic](https://github.com/khoih-prog/WebSockets_Generic) library

![Selection_171](https://user-images.githubusercontent.com/57012152/197635114-657d2995-3888-4bd4-b684-4a40ff0ca2b0.png)

**Comment by salrodri** [comment]
I am using an esp32.
So I just compile it and upload it and it seems to work.
I will go ahead and turn off the wifi on my router and see if it works.
[image: image.png]
